discourse_check_cleanup.py

```diff
@@ -0,0 +1,348 @@
+import contextlib
+import json
+import logging
+import sys
+from enum import Enum
+
+from src.discourse import Discourse, create_discourse
+from src.exceptions import DiscourseError
+
+
+class Action(str, Enum):
+    """The actions the utility can take.
+
+    Attrs:
+        CHECK_DRAFT: Check that the draft integration test succeeded.
+        CHECK_CREATE: Check that the create integration test succeeded.
+        CHECK_DELETE_TOPICS: Check that the delete_topics integration test succeeded.
+        CHECK_DELETE: Check that the delete integration test succeeded.
+        CLEANUP: Discourse cleanup after the testing.
+    """
+
+    CHECK_DRAFT = "check-draft"
+    CHECK_CREATE = "check-create"
+    CHECK_DELETE_TOPICS = "check-delete-topics"
+    CHECK_DELETE = "check-delete"
+    CLEANUP = "cleanup"
+
+
+def main():
+    """Clean up created Discourse pages."""
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser(
+        prog="DiscourseCleanup",
+        description="Delete posts created on discourse during an action execution.",
+    )
+    parser.add_argument("urls_with_actions", help="The pages that were created during execution")
+    parser.add_argument(
+        "discourse_config", help="The discourse configuration used to create the pages"
+    )
+    parser.add_argument(
+        "--action", help="Action to run", choices=tuple(action.value for action in Action)
+    )
+    parser.add_argument("--action-kwargs", help="Arguments for the action as a JSON mapping")
+    args = parser.parse_args()
+    urls_with_actions = json.loads(args.urls_with_actions)
+    discourse_config = json.loads(args.discourse_config)
+    action_kwargs = json.loads(args.action_kwargs or "{}")
+
+    discourse = create_discourse(**discourse_config)
+
+    match args.action:
+        case Action.CHECK_DRAFT.value:
+            _exit_with_result(check_draft(urls_with_actions=urls_with_actions, **action_kwargs))
+        case Action.CHECK_CREATE.value:
+            _exit_with_result(
+                check_create(
+                    urls_with_actions=urls_with_actions, discourse=discourse, **action_kwargs
+                )
+            )
+        case Action.CHECK_DELETE_TOPICS.value:
+            _exit_with_result(
+                check_delete_topics(
+                    urls_with_actions=urls_with_actions, discourse=discourse, **action_kwargs
+                )
+            )
+        case Action.CHECK_DELETE.value:
+            _exit_with_result(
+                check_delete(
+                    urls_with_actions=urls_with_actions, discourse=discourse, **action_kwargs
+                )
+            )
+        case Action.CLEANUP.value:
+            cleanup(urls_with_actions=urls_with_actions, discourse=discourse, **action_kwargs)
+            sys.exit(0)
+        case _:
+            raise NotImplementedError(f"{args.action} has not been implemented")
+
+
+def _exit_with_result(check_result: bool) -> None:
```

Comment:
I'd rather throw an exception is an error is detected and the catch it and exit as appropriate instead of controlling the exceptional flow using booleans. You can just raise or re-raise the exception and capture and log at the end of the `match`block.

Reply:
This utility can be best thought of as a linter. The various `check_...` functions are essentially performing checks similar to linters. A failing check is not necessarily an exception in the flow of the utility as it does not require the check to pass. The `exit_with_code` function is designed to provide an appropriate exit code for GitHub actions to know whether the checks passed or failed, similar to that a tox command could pass or fail. Does that make sense?

Reply:
It does